Global sync phase control

moxcodes opened the pull request:
## Proposed changes

I'm afraid the dependencies for this are starting to stack pretty high, but because this PR involves some delicate decisions about the core parallelization utilities, I'd open this and see if anyone had feedback on how we should go about accomplishing these global syncs that we'll use for checkpointing and load-balancing (and likely other future applications, like AMR).

### Types of changes:

- [ ] Bugfix
- [ ] New feature
- [ ] Refactor

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Dependencies

- [x] #2324
- [x] #2368 
- [x] #2374

moxcodes committed:
Add phase selection entry methods to Main chare

moxcodes committed:
Add global sync phase control to algorithm

fmahebert commented on src/Parallel/Main.hpp:
```diff
@@ -66,9 +66,15 @@ class Main : public CBase_Main<Metavariables> {
+  /// next requested phase will be run.
   void execute_next_phase() noexcept;
 
+  /// Request a phase for future execution during global syncs.
```
- To clarify that requests are only valid once, I suggest "for future execution during global syncs" -> "to execute after the next global sync"

- To clarify that multiple phases can be requested, I suggest "a phase" -> "one or more phases" here, and change things to plural below (at least argument -> `phase_requests`, maybe also function name -> `request_next_phases` but I'm not as sure about that one).

... assuming I correctly parsed the code, that is.

fmahebert commented on src/Parallel/Main.hpp:
```diff
@@ -87,6 +93,8 @@ class Main : public CBase_Main<Metavariables> {
 
   CProxy_ConstGlobalCache<Metavariables> const_global_cache_proxy_;
+  std::set<typename Metavariables::Phase> requested_phases_;
+  boost::optional<typename Metavariables::Phase> return_phase_;
```
Do I correctly understand this to be, e.g., the evolution phase in the context of "evolution -> checkpoint/LB -> evolution"? I find the name a little confusing, because "return" has so much meaning already in C++. Here's some alternate options I'm not entirely happy with: `return_to_this_phase_`, `phase_to_return_to_`, `phase_to_resume_`... what do you think?

fmahebert commented on src/Parallel/Main.hpp:
```diff
@@ -384,8 +392,18 @@ void Main<Metavariables>::
+  } else {
+    current_phase_ = *return_phase_;
+    return_phase_ = boost::none;
+  }
```
[nitpick] I think I'd find this slightly easier to parse as two-deep nested ifs, rather than three if/else cases. But this is super trivial, feel free to disagree.

fmahebert commented on src/Parallel/Algorithm.hpp:
```diff
@@ -30,6 +30,7 @@
 #include "Parallel/CharmRegistration.hpp"
 #include "Parallel/ConstGlobalCache.hpp"
 #include "Parallel/Info.hpp"
+#include "Parallel/Main.decl.h"
```
I think we're mostly (but not always) putting the charm includes after our standard includes. I don't think we ask this in the guidelines, but I do like making these special includes stand out... what are your thoughts on this?

moxcodes replied:
I think it does make sense to include them in a separate block following the standard includes -- I'll do that if it doesn't cause some problem with include order

fmahebert commented on src/Parallel/Algorithm.hpp:
```diff
@@ -331,31 +401,92 @@ class AlgorithmImpl<ParallelComponent, tmpl::list<PhaseDepActionListsPack...>>
 
   /// Tell the Algorithm it should no longer execute the algorithm. This does
   /// not mean that the execution of the program is terminated, but only that
   /// the algorithm has terminated. An algorithm can be restarted by pass `true`
```
Not your fault, but a typo in this documentation: pass -> passing

fmahebert commented on src/Parallel/Algorithm.hpp:
```diff
@@ -331,31 +401,92 @@ class AlgorithmImpl<ParallelComponent, tmpl::list<PhaseDepActionListsPack...>>
+          number_of_actions_in_phase(phase_) != 0;
+    }
+    if (should_start_phase) {
+          if (not get_terminate() and not wait_until_next_phase_) {
```
Apply clang-format

fmahebert commented on src/Parallel/Algorithm.hpp:
```diff
@@ -409,6 +540,66 @@ class AlgorithmImpl<ParallelComponent, tmpl::list<PhaseDepActionListsPack...>>
     return number_of_actions;
   }
 
+  // Invoke the action's static `apply` method. The if constexprs are for
```
I suggest increased clarity via "Invoke the static `apply` method on `ThisAction`", or similar.

fmahebert commented on src/Parallel/Algorithm.hpp:
```diff
@@ -409,6 +540,66 @@ class AlgorithmImpl<ParallelComponent, tmpl::list<PhaseDepActionListsPack...>>
 
+  // Invoke the action's static `apply` method. The if constexprs are for
+  // handling the cases where the `apply` method returns a tuple of one, two,
+  // three, or four elements, in order:
```
A discussion of this new fourth element should be added to the parallelization dev guide.

fmahebert commented on src/Parallel/Algorithm.hpp:
```diff
@@ -409,6 +540,66 @@ class AlgorithmImpl<ParallelComponent, tmpl::list<PhaseDepActionListsPack...>>
+  //    during the next global sync.
+  //
+  // The first argument to the invokable is the DataBox to be passed into
+  // the action's `apply` method, while the second is:
```
Two confusing things in this,
- what is "the invokable"?
- what second argument?

I can't tell if this is an outdated snippet or if I'm failing to apply it to the code.. help?

moxcodes replied:
oops yep this is outdated, from the comment that preceded the lambda I replaced and I didn't adequately fix the comment. I think this part can now just be removed.

fmahebert commented on src/Parallel/Algorithm.hpp:
```diff
@@ -409,6 +540,66 @@ class AlgorithmImpl<ParallelComponent, tmpl::list<PhaseDepActionListsPack...>>
+        "The return type must be a tuple of length one, two, three, or four "
+        "with:"
+        " first type is an updated DataBox;"
+        " second type is either a bool (indicating termination);"
```
"is either a" -> "is a" ?

fmahebert commented on src/Parallel/Algorithm.hpp:
```diff
@@ -409,6 +540,66 @@ class AlgorithmImpl<ParallelComponent, tmpl::list<PhaseDepActionListsPack...>>
+      algorithm_step_ = get<2>(action_return);
+    }
+    if constexpr (tuple_size >= 4_st) {
+      auto& phases = get<3>(action_return);
```
can this be const?

fmahebert commented on src/Parallel/Algorithm.hpp:
```diff
@@ -409,6 +540,66 @@ class AlgorithmImpl<ParallelComponent, tmpl::list<PhaseDepActionListsPack...>>
+    }
+    if constexpr (tuple_size >= 4_st) {
+      auto& phases = get<3>(action_return);
+      for (auto phase : phases) {
```
can this be const? perhaps a reference?

fmahebert commented on src/Parallel/Algorithm.hpp:
```diff
@@ -105,7 +107,71 @@ template <typename Component>
+
+template <typename ParallelComponentType, typename ActionType,
+          typename GeneralType>
+struct check_iterable_action_return_type : std::false_type {};
```
For my understanding: why does this have the `ActionType` template parameter? Unlike the other two parameters, I don't see this one being used. Could it be removed?

moxcodes replied:
You are correct that it is not used, but I've added the template parameter so that when the `static_assert` fails, the action that has a bad return type is printed. I'll add a code comment explaining that.

fmahebert commented on src/Parallel/Algorithm.hpp:
```diff
@@ -331,31 +401,92 @@ class AlgorithmImpl<ParallelComponent, tmpl::list<PhaseDepActionListsPack...>>
+    bool should_start_phase = false;
+    if constexpr(detail::has_required_phase_sync_function<metavariables>) {
+      should_start_phase =
+          (not static_cast<bool>(return_phase_) or phase_ == *return_phase_ or
```
I'm confused by the logic here (and in else branch below).

I was expecting the bool cast of `return_phase_` to control only the evaluation of the check `phase == return_phase_`, as on line 429 below. But if I'm reading the code here correctly, instead this is saying that an unset `return_phase_` should start the next phase, which I don't grasp the meaning of.

If this isn't a bug, please add a comment explaining? If this is a bug, and the logic should be as in line 429, perhaps the "set and equal" statement of 429 could be assigned to a variable for reuse in these multiple lines.

moxcodes replied:
So, the `return_phase_` only gets set when the algorithm is told to stop for the sync; then it represents the phase that will be returned to after the collection of 'sync phases'. So, if the return phase is unset, we are not currently going through the set of sync phases, so we should proceed through the phase list normally (so it accepts the next phase dictated by the main chare).
I agree the logic here can be pretty confusing because of all of the state possibilities when it starts a new phase, so I'll try to clarify in the comments, but perhaps you have some ideas about how this can be clearer with better variable names and documentation elsewhere?

fmahebert commented on src/Parallel/Algorithm.hpp:
```diff
@@ -317,7 +386,8 @@ class AlgorithmImpl<ParallelComponent, tmpl::list<PhaseDepActionListsPack...>>
   /// Start evaluating the algorithm until the is_ready function of an Action
-  /// returns false, or an Action returns with `terminate` set to `true`
+  /// returns false, or an Action returns with `terminate` set to `true` or
+  /// returns a different phase to be executed next.
```
Big picture discussion: if an action says "next time we get to a global sync, we'll need to checkpoint", why should that interrupt the flow of the algorithm?

Do I misunderstand what you wrote here? I thought we wanted (in this first-pass implementation) not to have any requests that interrupt flow.

moxcodes replied:
You're correct -- I just failed to correct this comment after one of the code iterations, so I'll adjust it to reflect the current state.

fmahebert commented on src/Parallel/Algorithm.hpp:
```diff
@@ -331,31 +401,92 @@ class AlgorithmImpl<ParallelComponent, tmpl::list<PhaseDepActionListsPack...>>
+  ///
+  /// The global sync points are identified by calling `stop_for_sync_phases` at
+  /// a common point in the execution across elements (e.g. a slab boundary for
+  /// slab index set in the input file). During those global sync points,
```
I thought global sync points would be given by a user-provided frequency in input file. I'm not sure what you mean when you say they're identified by calling `stop_for_sync_phases`. Can you try rephrasing this sentence?

moxcodes replied:
So, that's correct, but the way I'm planning to get that information from the input file is to have a special trigger which will ultimately call this charm entry function. I'll try to rephrase in a more user-facing form.

fmahebert commented on src/Parallel/Algorithm.hpp:
```diff
@@ -409,6 +540,66 @@ class AlgorithmImpl<ParallelComponent, tmpl::list<PhaseDepActionListsPack...>>
+  // to the next phase)
+  // 3. an unsigned integer corresponding to which action in the current phase's
+  //    algorithm to execute next.
+  // 4. a std::vector<metavariables::Phase> of phases to request to jump to
```
It's annoying me that an action provides the requests as a vector, which we first convert to an unordered_set, and then into a set. Can you convince me this is reasonable? I'd otherwise suggest making the actions provide their requests as an unordered_set, too.

moxcodes replied:
I'm happy to have the actions send `unordered_set`s

fmahebert commented on src/Parallel/Algorithm.hpp:
```diff
@@ -423,6 +614,9 @@ class AlgorithmImpl<ParallelComponent, tmpl::list<PhaseDepActionListsPack...>>
 
+  std::unordered_set<PhaseType> my_sync_phases_{};
+  boost::optional<PhaseType> return_phase_;
+  bool wait_until_next_phase_{false};
```
General complaint, though I've alluded to this in an earlier comment: I'm finding the naming confusing. I think sometimes you're trying to use "sync phases" so denote the "inserted" phases at global syncs? I think this would be a fine approach if done more consistently. As an alternative, here are some possible renames...

- `my_sync_phases_` -> `requested_phases_after_sync_` or `requested_next_phases_` or ...
- `return_phase_` -> `phase_to_return_to_` or `phase_to_resume_` or ...
- `wait_until_next_phase_` -> `sleep_algorithm_until_next_phase_` or similar (or will this clash with a hopefully-coming rename of "terminate"?)

I'm definitely not attached to these names, and would be happy to discuss in more detail to help converge on something.

fmahebert commented on src/Parallel/Algorithm.hpp:
```diff
@@ -720,112 +879,96 @@ AlgorithmImpl<ParallelComponent, tmpl::list<PhaseDepActionListsPack...>>::
-              } else {
-                display_databox_error(__LINE__);
-              }
+          // clang-format on
```
Uh oh, do I correctly see that most of this diff is formatting? Can I be a pain and request you move the whitespace changes to a separate commit? (and below)

moxcodes replied:
I've done one better and entirely removed all of the make_overloader business from this file, converting to far more succinct `if constexpr`s.

fmahebert reviewed (commented):
Here's a partial review; I haven't looked at the test yet, but I'll get to that next week. If these comments reveal a lack of general understanding on my part, let's talk on Monday/Tuesday :)

moxcodes committed:
Document global sync phases in dev guide

moxcodes commented:
Thanks for the review @fmahebert ! I've posted some fixups -- the changes to `Algorithm.hpp` were sufficiently large that I just squashed them in to the `Add global sync phase control to algorithm`. I also added a documentation commit. (sorry for the duplicate, I posted the first one on the wrong PR)

moxcodes committed:
Add Main proxy to GlobalCache

moxcodes committed:
Add serialization functions to Algorithm objects

- These should likely inherit from their base class, but
is probably better practice to keep them defined as suggested
in charm documentation

fmahebert commented on src/Parallel/GlobalCache.ci:
```diff
@@ -4,6 +4,7 @@
 module GlobalCache {
   include "Parallel/ParallelComponentHelpers.hpp";
   include "Utilities/TaggedTuple.hpp";
+  include "Parallel/Main.decl.h";
```
Commit message: ConstGlobalCache -> GlobalCache

fmahebert commented on src/Parallel/GlobalCache.hpp:
```diff
@@ -372,6 +374,14 @@ class GlobalCache : public CBase_GlobalCache<Metavariables> {
+  /// called once.
+  void set_main_proxy(const main_proxy_type& main_proxy) noexcept;
+
+  /// Retrieve the proxy to the Main chare (or boost::none if the proxy has not
```
boost::none -> std::nullopt

fmahebert commented on src/Parallel/GlobalCache.hpp:
```diff
@@ -487,6 +498,23 @@ GlobalCache<Metavariables>::get_this_proxy() noexcept {
+template <typename Metavariables>
+void GlobalCache<Metavariables>::set_main_proxy(
+    const main_proxy_type& main_proxy) noexcept {
+  if (not static_cast<bool>(main_proxy_)) {
```
Prefer `main_proxy_.has_value()` instead of casting

fmahebert commented on src/Parallel/Main.hpp:
```diff
@@ -264,7 +274,7 @@ Main<Metavariables>::Main(CkArgMsg* msg) noexcept {
                                                    const_global_cache_tags{}),
       mutable_global_cache_proxy_, &mutable_global_cache_dependency);
-  const_global_cache_proxy_.set_main_proxy(this->thisProxy);
+  global_cache_proxy_.set_main_proxy(this->thisProxy);
```
I think this rename should have been in the previous commit?

fmahebert commented on src/Parallel/Main.hpp:
```diff
@@ -393,8 +403,21 @@ void Main<Metavariables>::
 void Main<Metavariables>::execute_next_phase() noexcept {
-  current_phase_ = Metavariables::determine_next_phase(
-      current_phase_, global_cache_proxy_);
+  if (not static_cast<bool>(phase_to_resume_after_sync_phases_)) {
```
`static_cast` -> `.has_value()`

fmahebert commented on src/Parallel/Algorithm.hpp:
```diff
@@ -55,6 +57,152 @@ template <typename ParallelComponent, typename PhaseDepActionList>
+template <typename Component>
+using get_charm_base_class_t =
+    typename get_charm_base_class<Component,
+                                  typename Component::chare_type>::type;
```
I don't see `get_charm_base_class` being used elsewhere in the PR. Is it needed?

Furthermore, isn't this information now encapsulated by `Component::cbase` (if I've correctly understood that Component is one of the structs in Parallel::Algorithms?)?

moxcodes replied:
ah, shoot -- yeah, you're right, this is a change that my `serialize_algorithm` alterations means this entire type trait should just be removed.

fmahebert commented on src/Parallel/Algorithm.hpp:
```diff
@@ -281,31 +434,90 @@ class AlgorithmImpl<ParallelComponent, tmpl::list<PhaseDepActionListsPack...>>
+    bool should_start_phase = false;
+    if constexpr(detail::has_required_phase_sync_function<metavariables>) {
+      should_start_phase =
+          (not static_cast<bool>(phase_to_resume_after_sync_phases_) or
```
`static_cast` -> `has_value()`

... here and elsewhere in this commit.

fmahebert reviewed (commented):
A quick pass with fairly minor stylistic complaints. You can squash these directly... I'll take another deeper look later

moxcodes commented:
squashed in a fix to the test CMakeLists where it wasn't actually running the test on CI and a fix to the `RestartMe` action in the test where it was restarting the wrong component, but was also being run on the wrong component, so with a test with only two components the mistakes cancelled out.

kidder commented on src/Parallel/Main.hpp:
```diff
@@ -264,6 +264,7 @@ Main<Metavariables>::Main(CkArgMsg* msg) noexcept {
       Parallel::create_from_options<Metavariables>(options_,
                                                    const_global_cache_tags{}),
       mutable_global_cache_proxy_, &mutable_global_cache_dependency);
+  global_cache_proxy_.set_main_proxy(this->thisProxy);
```
why not just pass the main proxy to the constructor of the global_cache_proxy ?  Then you don't need the set_main_proxy entry method

moxcodes replied:
Yes, I agree, that'll be better

kidder reviewed (changes requested):
one additional comment besides what we discussed on the hangout

moxcodes commented:
I'm going to go ahead and close this. Partially superseded by #2892 and #2960, all the rest will need to be rewritten for other changes.
